### PR TITLE
expression: convert charset by wrapping internal builtin function

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Problem Summary: Close https://github.com/pingcap/tidb/issues/29863.

This is a refactoring PR. **The code changes about built-in functions in TiKV can be minimized.**

After the investigation about several built-in functions, I found some common patterns:

1. Create a new encoding.
1. Convert the arguments into new charset from utf8.
1. Apply these changes to the vectorized-version evaluation.

We can extract these patterns into another `Sig` named `builtinToBinarySig`(`to_binary()`) and wrap the cast function as follows:

The rewrite process **before** this PR:
```sql
select some_builtin(arg1, arg2...)
```
is rewritten to 
```sql
select some_builtin(cast_as_string(arg1), cast_as_string(arg2)...)
```

The rewrite process **after** this PR:
```sql
select some_builtin(arg1, arg2...)
```
is rewritten to 
```sql
select some_builtin(to_binary(cast_as_string(arg1)), to_binary(cast_as_string(arg2))...)
```

Then we can maintain a built-in function list to determine which function needs to be wrapped, instead of putting the boilerplates everywhere.

```go
var toBinaryMap = map[string]struct{}{
	ast.Hex: {}, ast.Length: {}, ast.OctetLength: {}, ast.ASCII: {},
	ast.ToBase64: {},
}
```

### What is changed and how it works?

See the code.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1907,6 +1908,33 @@ func WrapWithCastAsDecimal(ctx sessionctx.Context, expr Expression) Expression {
 	return BuildCastFunction(ctx, expr, tp)
 }
 
+// charsetConvertMap contains the builtin functions which arguments need to be converted to the correct charset.
+var charsetConvertMap = map[string]struct{}{
+	ast.Hex: {}, ast.Length: {}, ast.OctetLength: {}, ast.ASCII: {},
```
Comment (Defined2014) [inline expression/builtin_cast.go:1913]: Maybe other functions can also be added here, like md5(), sha1()... 

reset, LGTM

Comment (tangenta) [reply expression/builtin_cast.go:1913]: Yea. We need to check them one by one later.

File: expression/builtin_convert_charset.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	ast.Hex: {}, ast.Length: {}, ast.OctetLength: {}, ast.ASCII: {},
+	ast.ToBase64: {},
+}
+
+// WrapWithToBinary wraps `expr` with to_binary sig.
+func WrapWithToBinary(ctx sessionctx.Context, expr Expression, funcName string) Expression {
+	retTp := expr.GetType()
+	if _, err := charset.GetDefaultCollationLegacy(retTp.Charset); err != nil {
+		if _, ok := toBinaryMap[funcName]; ok {
+			bf, err := newBaseBuiltinFunc(ctx, InternalFuncToBinary, []Expression{expr}, retTp.EvalType())
```
Comment (xiongjiwei) [inline expression/builtin_convert_charset.go:125]: maybe we can use `getFunction`, it will also set pbcode

Comment (tangenta) [reply expression/builtin_convert_charset.go]: Done.

**Review by xiongjiwei — APPROVED** [review]

File: expression/builtin_convert_charset.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	argFieldTp := args[0].GetType()
+	argTp := argFieldTp.EvalType()
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, argTp, argTp)
+	if err != nil {
+		return nil, err
+	}
+	var sig builtinFunc
+	switch argTp {
+	case types.ETString:
```
Comment (xiongjiwei) [inline expression/builtin_convert_charset.go:48]: why not check the argTp earlier

File: expression/builtin_convert_charset.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	baseFunctionClass
+}
+
+func (c *tidbConvertCharsetFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, c.verifyArgs(args)
+	}
+	argFieldTp := args[0].GetType()
+	argTp := argFieldTp.EvalType()
+	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, argTp, argTp)
```
Comment (xiongjiwei) [inline expression/builtin_convert_charset.go:42]: then we can 
```suggestion
	bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString)
```

File: expression/builtin_convert_charset.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	if err != nil {
+		return nil, err
+	}
+	var sig builtinFunc
+	switch argTp {
+	case types.ETString:
+		sig = &builtinInternalToBinarySig{bf}
+		// TODO(tangenta): set pbcode for to_binary() sig.
```
Comment (xiongjiwei) [inline expression/builtin_convert_charset.go:50]: this todo can be done

File: expression/builtin_convert_charset.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+			return err
+		}
+		result.AppendString(str)
+	}
+	return nil
+}
+
+// toBinaryMap contains the builtin functions which arguments need to be converted to the correct charset.
+var toBinaryMap = map[string]struct{}{
+	ast.Hex: {}, ast.Length: {}, ast.OctetLength: {}, ast.ASCII: {},
```
Comment (zimulala) [inline expression/builtin_convert_charset.go:115]: How about other functions like `upper`？

Comment (tangenta) [reply expression/builtin_convert_charset.go:113]: Let's add them later.

File: expression/builtin_convert_charset.go (hunk 3 above)
Comment (tangenta) [reply expression/builtin_convert_charset.go:43]: Updated.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**